exec/java-exec/src/main/java/org/apache/drill/exec/planner/fragment/SimpleParallelizerMultiPlans.java

```diff
@@ -0,0 +1,222 @@
+      Collection<DrillbitEndpoint> activeEndpoints, PhysicalPlanReader reader, Fragment rootFragment,
+      UserSession session, QueryContextInformation queryContextInfo) throws ExecutionSetupException {
+
+    final PlanningSet planningSet = getFragmentsHelper(activeEndpoints, rootFragment);
+
+    return generateWorkUnits(
+        options, foremanNode, queryId, reader, rootFragment, planningSet, session, queryContextInfo);
+  }
+
+  /**
+   * Split plan into multiple plans based on parallelization
+   * Ideally it is applicable only to plans with two major fragments: Screen and UnionExchange
+   * But there could be cases where we can remove even multiple exchanges like in case of "order by"
+   * End goal is to get single major fragment: Screen with chain that ends up with a single minor fragment
+   * from Leaf Exchange. This way each plan can run independently without any exchange involvement
+   * @param options
+   * @param foremanNode - not really applicable
+   * @param queryId
+   * @param reader
+   * @param rootNode
+   * @param planningSet
+   * @param session
+   * @param queryContextInfo
+   * @return
+   * @throws ExecutionSetupException
+   */
+  private List<QueryWorkUnit> generateWorkUnits(OptionList options, DrillbitEndpoint foremanNode, QueryId queryId,
+      PhysicalPlanReader reader, Fragment rootNode, PlanningSet planningSet,
+      UserSession session, QueryContextInformation queryContextInfo) throws ExecutionSetupException {
+
+    // now we generate all the individual plan fragments and associated assignments. Note, we need all endpoints
+    // assigned before we can materialize, so we start a new loop here rather than utilizing the previous one.
+
+    List<QueryWorkUnit> workUnits = Lists.newArrayList();
+    int plansCount = 0;
+    DrillbitEndpoint[] endPoints = null;
+    long initialAllocation = 0;
+    long maxAllocation = 0;
+
+    final Iterator<Wrapper> iter = planningSet.iterator();
+    while (iter.hasNext()) {
+      Wrapper wrapper = iter.next();
+      Fragment node = wrapper.getNode();
+      boolean isLeafFragment = node.getReceivingExchangePairs().size() == 0;
+      final PhysicalOperator physicalOperatorRoot = node.getRoot();
+      // get all the needed info from leaf fragment
+      if ( (physicalOperatorRoot instanceof Exchange) &&  isLeafFragment) {
+        // need to get info about
+        // number of minor fragments
+        // assignedEndPoints
+        // allocation
+        plansCount = wrapper.getWidth();
+        initialAllocation = (wrapper.getInitialAllocation() != 0 ) ? wrapper.getInitialAllocation()/plansCount : 0;
+        maxAllocation = (wrapper.getMaxAllocation() != 0 ) ? wrapper.getMaxAllocation()/plansCount : 0;
+        endPoints = new DrillbitEndpoint[plansCount];
+        for (int mfId = 0; mfId < plansCount; mfId++) {
+          endPoints[mfId] = wrapper.getAssignedEndpoint(mfId);
```

Comment:
Since the end-point assignments are saved with the plan fragment which will be submitted later (maybe after several minutes or even hours depending on the client), it is possible that due to data movement (e.g rebalancing) the data locality could have changed in the meantime.  Is there additional check done at the time the split plan is run ? Is that a concern ?